### ISSUE TITLE
[sharding] shard transfer - receiving end 

### ISSUE THREAD

**Issue by agourlay** [qdrant/qdrant#789]
**Is your feature request related to a problem? Please describe.**
We need a way for nodes to receive shard transfer.

**Describe the solution you'd like**
Three new internal gRPC endpoints:

1. Create temporary shard structure
2. Receive batched records into temporary shard
3. End of transfer - promote temporary shard into regular local shard

Requirements:

1. The temporary shard must not interfere with existing shards
2. The temporary shard must survive a node restart  

**Describe alternatives you've considered**
Receiving node could somehow pull the data but I am not sure it would be better.

**Comment by generall** [comment]
> Receive batched records into temporary shard

We can reuse existing update API for this, and rout internally by the shard id.

We would require a dedicated API for file-based transfer, but it can wait for another task

**Comment by agourlay** [comment]
Got a discussion with @e-ivkov (thanks!) about the best way to design the temporary shards to smooth-en the interaction with the consensus & update API.

We want to avoid:
- having to manipulate the shard mapping before the end of the transfer to avoid data corruption
- having to manipulate the ring with a temporary shardID

We propose to add a field containing the current transfer state at the collection level which will contain:

- the list of ongoing transfer impacting this collection
- the mapping from `shardId` to temporary shard

A temporary shard is a special local shard that lives in a different directory to not interfere with the current collection's data.

When an upsert points API call is received for a specific `shardID`, we need to handle the case where the `shardID` resolves to a remote shard and do an additional lookup to find if there is a corresponding temporary shard.

The goal is to reuse as much as the update infrastructure as possible.

Once the transfer is done, we need to promote atomically the temporary shard:

- move the data in the collection/shards folder
- update collection's shards mapping in place to replace the remote shard by a now local shard properly load
